metatomic-torch/src/io/io.cpp

```diff
@@ -0,0 +1,244 @@
+    zw.add_file("cell.npy", bytes, io::ZIP_STORED);
+  }
+  {
+    auto t = system->types().contiguous().cpu(); // keep i32 or i64
+    auto bytes = io::npy_write(t);
+    zw.add_file("types.npy", bytes, io::ZIP_STORED);
+  }
+  {
+    auto bytes = io::npy_write(system->pbc().contiguous().to(torch::kBool).cpu());
+    zw.add_file("pbc.npy", bytes, io::ZIP_STORED);
+  }
+
+  // Neighbor lists
+  {
+    auto nls = system->known_neighbor_lists(); // std::vector<NeighborListOptions>
+    for (size_t i = 0; i < nls.size(); ++i) {
+      const auto& opts = nls[i];
+      const std::string base = "pairs/" + std::to_string(i) + "/";
+
+      // options.json (JSON string)
+      std::string json = opts->to_json();
+      zw.add_file(base + "options.json",
+                  reinterpret_cast<const uint8_t*>(json.data()), json.size(), 0);
+
+      // data.mts (TensorBlock bytes via member save_buffer)
+      TensorBlock block = system->get_neighbor_list(opts);
+      torch::Tensor mts_buf = metatensor_torch::save_buffer(block);
+      mts_buf = mts_buf.contiguous().cpu();
+      zw.add_file(base + "data.mts",
+                  mts_buf.data_ptr(), static_cast<size_t>(mts_buf.numel()), 0);
+    }
+  }
+
+  // Extra data (TensorMap)
+  {
+    auto keys = system->known_data(); // std::vector<std::string>
+    for (const auto& key : keys) {
+      TensorMap tmap = system->get_data(key);
+      torch::Tensor mts_buf = metatensor_torch::save_buffer(tmap);
+      mts_buf = mts_buf.contiguous().cpu();
+      zw.add_file("data/" + key + ".mts",
+                  mts_buf.data_ptr(), static_cast<size_t>(mts_buf.numel()), 0);
+    }
+  }
+}
+
+void save(const std::string& path, const System& system) {
+  require_mta_extension(path);
+  io::ZipWriter zw;
+  zw.open_file(path, /*flags*/ 0);
+  write_system_to_zip(zw, system);
+  zw.finalize();
+}
+
+torch::Tensor save_buffer(const System& system) {
+  io::ZipWriter zw;
+  zw.open_memory(/*initial*/ 0, /*flags*/ 0);
+  write_system_to_zip(zw, system);
+  auto bytes = zw.finalize_to_vector();
+
+  // create a tensor and copy bytes in
+  auto bytes_as_tensor = torch::empty(
+      { static_cast<long>(bytes.size()) },
+      torch::TensorOptions().dtype(torch::kUInt8).device(torch::kCPU)
+  );
+  if (!bytes.empty()) {
+      std::memcpy(bytes_as_tensor.data_ptr<uint8_t>(), bytes.data(), bytes.size());
+  }
+  return bytes_as_tensor;
+}
+
+// ---------- read: ZIP -> System (file/memory)
+
+static System read_system_from_zip(io::ZipReader& zr) {
+  using metatensor_torch::TensorBlock;
+  using metatensor_torch::TensorMap;
+
+  // Validate required files exist
+  ensure(zr.has("positions.npy") && zr.has("cell.npy") &&
+         zr.has("types.npy") && zr.has("pbc.npy"),
+         "File does not contain a valid System object (.npy core missing)");
+
+  // Load core arrays
+  auto pos   = io::npy_read(zr.read("positions.npy")).contiguous().to(torch::kFloat64).cpu();
```

Comment:
This should not need any cast; the returned data is always contiguous and on CPU. 

Same below